tools/src/main/scala/org/apache/spark/tools/GenerateMIMAIgnore.scala

```diff
@@ -80,20 +79,25 @@ object GenerateMIMAIgnore {
         /* Inner classes defined within a private[spark] class or object are effectively
          invisible, so we account for them as package private. */
         lazy val indirectlyPrivateSpark = {
-          val maybeOuter = className.toString.takeWhile(_ != '$')
```

Comment:
This old code didn't behave so nicely in cases where `className` was an un-nested companion object. For example, say you have a class named `org.apache.spark.HeartbeatReceiver$`: this class isn't an inner class and the old code would have `maybeOuter = "org.apache.spark.HeartbeatReceiver"`, which might not be a valid class.

The replacement code that I added isn't perfect but fails far less often.